feat: added Bitcoin Contracts entry point and list page

Polybius93 opened the pull request:
> Try out this version of Leather — [download extension builds](https://github.com/leather-wallet/extension/actions/runs/6561150311).<!-- Sticky Header Marker -->

This PR incorporates the following changes:

- The retrieval mechanism for Bitcoin Contracts, previously absent, is now fully implemented.
- The asset list has been updated to include a new entry point for Bitcoin Contracts, displaying the deposited Bitcoin balance for user convenience.
- A comprehensive Bitcoin Contracts List page has been introduced, showcasing individual Bitcoin Contract items.

<img width="391" alt="Screenshot 2023-09-27 at 17 12 11" src="https://github.com/leather-wallet/extension/assets/99192647/1a138bcc-dc17-4087-9c63-0dcf99bda790">
<img width="393" alt="Screenshot 2023-09-27 at 17 02 24" src="https://github.com/leather-wallet/extension/assets/99192647/9b1aacc6-9afe-403d-9f19-e6194e888a82">
<img width="392" alt="Screenshot 2023-09-27 at 17 02 12" src="https://github.com/leather-wallet/extension/assets/99192647/f2c60905-0170-4d69-a03d-66b0f01261a6">

markmhendrickson commented:
@Polybius93 mind rebasing and providing updated screenshots with the new brand, then we can re-review here? 🙏

fbwoolf reviewed (approved):
This lgtm but needs to be rebased on rebrand changes.

Polybius93 commented:
> @Polybius93 mind rebasing and providing updated screenshots with the new brand, then we can re-review here? 🙏

Sure, I'll be working on it this week! Thank you!

Polybius93 commented on src/app/components/icons/bitcoin-contract-icon.tsx:
```diff
@@ -0,0 +1,73 @@
+export function BitcoinContractIcon() {
```
This was coming from us, but yeah it would be nice to have a new one, can I ask your help with that?
@fbwoolf @mica000

Polybius93 commented:
> @Polybius93 mind rebasing and providing updated screenshots with the new brand, then we can re-review here? 🙏

Hey! I've added the new screenshots to the description!

mica000 commented:
@Polybius93 Possible to make the icon comply with our branding guidelines? Currently it looks a quite big in comparison to all the Tokens. The guidelines can be found in here:
https://www.figma.com/file/2MLHeIeL6XPVi3Tc2DfFCr/%E2%9D%96-Leather-%E2%80%93-Design-System?type=design&node-id=5709-83818&mode=design&t=WGxzh6MCewu6Nhwj-4

Let me know if you need help with it!

Polybius93 commented:
> @Polybius93 Possible to make the icon comply with our branding guidelines? Currently it looks a quite big in comparison to all the Tokens. The guidelines can be found in here: https://www.figma.com/file/2MLHeIeL6XPVi3Tc2DfFCr/%E2%9D%96-Leather-%E2%80%93-Design-System?type=design&node-id=5709-83818&mode=design&t=WGxzh6MCewu6Nhwj-4
> 
> Let me know if you need help with it!

Hey! Sure thing!

Can you give access to this link for these addresses:
rok@dlc.link
dani@dlc.link

Thank you!

Polybius93 commented:
> @Polybius93 Possible to make the icon comply with our branding guidelines? Currently it looks a quite big in comparison to all the Tokens. The guidelines can be found in here: https://www.figma.com/file/2MLHeIeL6XPVi3Tc2DfFCr/%E2%9D%96-Leather-%E2%80%93-Design-System?type=design&node-id=5709-83818&mode=design&t=WGxzh6MCewu6Nhwj-4
> 
> Let me know if you need help with it!

@mica000 Reworked the icon according to the guide, I hope this works for you! 
<img width="389" alt="Screenshot 2023-10-03 at 17 46 16" src="https://github.com/leather-wallet/extension/assets/99192647/0215935e-4567-4c1f-9d54-44b555417d3d">
<img width="402" alt="Screenshot 2023-10-03 at 17 46 27" src="https://github.com/leather-wallet/extension/assets/99192647/ba15bea1-2c6a-4321-96c2-661e2f51021a">

markmhendrickson commented:
@kyranjamie @mica000 This is ready for final review

kyranjamie commented on src/app/common/hooks/use-bitcoin-contracts.ts:
```diff
@@ -215,6 +274,32 @@ export function useBitcoinContracts() {
+  ): BitcoinContractListItem[] {
+    return bitcoinContracts.map(bitcoinContract => {
+      return {
+        id: truncateMiddle(bitcoinContract.id),
```
if this is just for presentation, wouldn't it be better to do this directly in the JSX where it's displayed, rather than in advance in the data structure?

Polybius93 replied:
Yeah, removed this formatting function

kyranjamie commented on src/app/components/bitcoin-contract-entry-point/bitcoin-contract-entry-point-layout.tsx:
```diff
@@ -0,0 +1,63 @@
+  isLoading?: boolean;
+  onClick: () => void;
+}
+export const BitcoinContractEntryPointLayout = forwardRefWithAs(
```
Feel free to leave this if you don't feel comfortable taking on yourself, but we're moving away from StacksUI in favour of Panda via `leather-styles/jsx`, which has a suite of similar components, `Flex`, `Box`, etc. 

Would be great to replace StacksUI with those

Polybius93 replied:
Removed it where it had the same attributes as in stacks/ui

kyranjamie commented on src/app/components/bitcoin-contract-entry-point/bitcoin-contract-entry-point.tsx:
```diff
@@ -0,0 +1,55 @@
+      setBitcoinContractSum(currentBitcoinContractSum);
+      setIsLoading(false);
+    };
+    getBitcoinContractDataAndSetState();
```
```suggestion
    void getBitcoinContractDataAndSetState();
```

removes warning

Polybius93 replied:
Thank you!

kyranjamie commented on src/app/pages/bitcoin-contract-list/bitcoin-contract-list.tsx:
```diff
@@ -0,0 +1,53 @@
+          return (
+            <BitcoinContractListItemLayout
+              key={bitcoinContract.id}
+              id={bitcoinContract.id}
```
Re-earlier comment

```suggestion
              id={truncateMiddle(bitcoinContract.id)}
```

this would make more sense to me as I wouldn't expect the data to be pre-formatted

Polybius93 replied:
Removed the format function, thank you for suggesting!

kyranjamie reviewed (commented):
Added some comments, @mica000 to review designs

mica000 commented:
@Polybius93 can you share the icon with me in a editable figma file we can help you adjusting?

Polybius93 commented:
> @Polybius93 can you share the icon with me in a editable figma file we can help you adjusting?

Hey! Sure, here you go!
https://www.figma.com/file/bT8x7wvBMvXOUlqNEk9Rpa/Leather-%2F-DLC.Link-Collab?type=design&node-id=0-1&mode=design&t=rgurb9qkIRz8iekt-0

mica000 commented:
@Polybius93 
I created another icon for you that is a bit smaller and its more consistent with ours. Also added to the Figma your shared.
![BitcoinLocked_Leather_ico_36px](https://github.com/leather-wallet/extension/assets/119523541/d12b53d5-d2f4-44f2-8de5-6fd66dc9700d)

Polybius93 commented:
> @Polybius93 I created another icon for you that is a bit smaller and its more consistent with ours. Also added to the Figma your shared. ![BitcoinLocked_Leather_ico_36px](https://user-images.githubusercontent.com/119523541/275777775-d12b53d5-d2f4-44f2-8de5-6fd66dc9700d.svg)

Thank you, @mica000! I've changed the icon as provided.

kyranjamie commented on src/app/query/bitcoin/contract/check-dlc-link-attestor-health.ts:
```diff
@@ -0,0 +1,17 @@
+export async function checkDlcLinkAttestorHealth(attestorUrl: string): Promise<boolean> {
+  let response: Response;
+
+  try {
+    response = await fetch(`${attestorUrl}/health`, {
+      method: 'get',
+    });
+  } catch (error) {
+    return false;
+  }
+
+  if (!response.ok) {
+    return false;
+  }
+
+  return true;
+}
```
Is it necessary to hoist the response var above and declare in the try/catch?

Isn't this the same as? 

```suggestion
  try {
    const response = await fetch(`${attestorUrl}/health`, {
      method: 'get',
    });
    
    return response.ok;
    
  } catch (error) {
    return false;
  }
}
```

Polybius93 replied:
You are right, I removed the unnecessary lines!

kyranjamie reviewed (approved):
LGTM thanks @Polybius93 

Just added a few last questions

Polybius93 commented on src/app/pages/bitcoin-contract-list/bitcoin-contract-list.tsx:
```diff
@@ -0,0 +1,52 @@
+  const [bitcoinContracts, setBitcoinContracts] = useState<BitcoinContractListItem[]>([]);
+  const [isLoading, setLoading] = useState(true);
+
+  useOnMount(() => {
+    const fetchAndFormatBitcoinContracts = async () => {
+      const fetchedBitcoinContracts = await getAllSignedBitcoinContracts();
+      setBitcoinContracts(fetchedBitcoinContracts);
+      setLoading(false);
+    };
+    void fetchAndFormatBitcoinContracts();
+  });
```
Currently, this fetching process is conducted through the @dlc-tools. We haven't planned to move it outside yet, as most of the contract handling and formatting occurs within the package. Is this a problem for you?
